### PR TITLE
Require `#[const_trait]` on `Trait` for `impl const Trait`

### PR DESCRIPTION
**Pull request by fee1-dead**
r? @oli-obk

### CODE REVIEW

File: compiler/rustc_passes/src/check_const.rs [hunk 1]
```diff
@@ -192,6 +192,28 @@ impl<'tcx> Visitor<'tcx> for CheckConstVisitor<'tcx> {
...
+                tcx.sess
+                    .struct_span_err(
+                        source_map.guess_head_span(item.span),
+                        "const `impl`s must be for traits marked with `#[const_trait]`",
+                    )
+                    .span_note(
+                        source_map.guess_head_span(tcx.def_span(def_id)),
+                        "this trait must be annotated with `#[const_trait]`",
+                    )
+                    .emit();
```
Comment (oli-obk) [inline compiler/rustc_passes/src/check_const.rs:214]: Add a test that shows these diagnostics

**Comment by oli-obk** [comment]
```
309 | static RT: Lazy<Mutex<Runtime>> = Lazy::new(Default::default);
    |                                   --------- ^^^^^^^^^^^^^^^^ the trait `~const Default` is not implemented for `Mutex<Runtime>`
    |                                   required by a bound introduced by this call
```

oh fun, adding `#[const_trait]` is a breaking change right now because all uses of the trait in const contexts will now assume we need `~const Trait`. This is somewhat why I really want constness to be a generic param, so we can have an `Infer` variant that will only fall back to the context after all other inference is done.

**Comment by fee1-dead** [comment]
> This is somewhat why I really want constness to be a generic param, so we can have an `Infer` variant that will only fall back to the context after all other inference is done.

Does that help? This error occurs in a static, so it is a const context. It just has to be fixed I think.

**Comment by oli-obk** [comment]
> This error occurs in a static, so it is a const context. It just has to be fixed I think.

yes, but inference would figure out that we require `NotConst` for the `Default::default` since the return type is `NotConst`. I don't quite know how to fix this in general. I mean users could be writing `let x = Default::default;` and at that point you have no idea whether you need const or not const and don't even know the other concrete generic params yet, so you don't know whether it can even be const. We need to delay figuring out the constness until a concrete type is known.

Hmm or are you saying we could not error and instead replace the obligation with a notconst one?

**Comment by fee1-dead** [comment]
`let x = Default::default;` wouldn't require the `Self` type to be const, right? Unless we are talking about const fn pointers

**Comment by oli-obk** [comment]
> `let x = Default::default;` shouldn't require the `Self` type to be const, right? Unless we are talking about const fn pointers

hmm good point, until we have those we can just always require notconst

File: compiler/rustc_passes/src/check_const.rs (hunk 1 above)
Comment (fee1-dead) [reply compiler/rustc_passes/src/check_const.rs:214]: Done.

**Comment by fee1-dead** [comment]
Checked performance locally, seems fine. Asking bors to check this again.

@bors try @rust-timer queue

**Comment by oli-obk** [comment]
The remaining perf regressions are really tiny in absolute time taken, so I think that would be fine.

File: src/test/ui/const-generics/issue-93647.stderr [hunk 2]
```diff
@@ -1,12 +1,17 @@
-error[E0015]: cannot call non-const closure in constants
+error[E0277]: the trait bound `[closure@$DIR/issue-93647.rs:2:6: 2:8]: Fn<()>` is not satisfied
```
Comment (oli-obk) [inline src/test/ui/const-generics/issue-93647.stderr:1]: This diagnostic is somewhat surprising. Why isn't it showing the `~const` when the `help` shows it?

Comment (fee1-dead) [reply src/test/ui/const-generics/issue-93647.stderr:1]: This is because `const_trait_impl` is not enabled.

Comment (oli-obk) [reply src/test/ui/const-generics/issue-93647.stderr:1]: I guess... still the main message is very confusing. We should probably figure out how to make it obvious to someone not in the know about the latest unstable const features

**Comment by fee1-dead** [comment]
```
curl: (22) The requested URL returned error: 429
```

Seems spurious.


@bors retry

**Comment by oli-obk** [comment]
```
error[E0277]: the trait bound `[closure@src/tools/clippy/tests/compile-test.rs:84:31: 84:33]: std::ops::Fn<()>` is not satisfied
  --> src/tools/clippy/tests/compile-test.rs:98:37
   |
98 |         if let Some((name, path)) = parse_name_path() {
   |                                     ^^^^^^^^^^^^^^^^^ expected an `Fn<()>` closure, found `[closure@src/tools/clippy/tests/compile-test.rs:84:31: 84:33]`
   |
   = help: the trait `~const std::ops::Fn<()>` is not implemented for closure `[closure@src/tools/clippy/tests/compile-test.rs:84:31: 84:33]`
note: the trait `std::ops::Fn<()>` is implemented for `[closure@src/tools/clippy/tests/compile-test.rs:84:31: 84:33]`, but that implementation is not `const`
  --> src/tools/clippy/tests/compile-test.rs:98:37
   |
98 |         if let Some((name, path)) = parse_name_path() {
   |                                     ^^^^^^^^^^^^^^^^^
   = note: wrap the `[closure@src/tools/clippy/tests/compile-test.rs:84:31: 84:33]` in a closure with no arguments: `|| { /* code */ }`
```

**Comment by fee1-dead** [comment]
Found a way to fix it. It also helped the error messages.

**Comment by fee1-dead** [comment]
Wait the log analyzer was wrong

@bors r=oli-obk

**Comment by Mark-Simulacrum** [comment]
> The remaining perf regressions are really tiny in absolute time taken, so I think that would be fine.

(from https://github.com/rust-lang/rust/pull/100982#issuecomment-1249480832)

@oli-obk, I feel like we've had this conversation before about absolute (extra) time taken being a terrible proxy for "fine" -- our benchmarks in general take very little time, so it's not a good measure of whether a given regression is significant. In this particular case, the regression to libc is small (~0.5%) but I believe real, likely caused by the additional calls to evaluate_obligation caused by this PR.

The _right_ justification here is that the additional calls are a necessary part of the feature, so we accept the additional overhead even for code not directly using it; we don't see opportunities for direct improvement of the PR's changes. (IOW, that the benefits here to end-users outweigh the negative performance effect on them). Obviously, this is just a proposal, not an actual justification, as I can't be confident in saying that there's not improvements that could be made here.

**Pull request merged**